### ISSUE TITLE
improve instance move performance

### ISSUE THREAD

**Issue by Ganeti-Issues-Migrator** [ganeti/ganeti#754]
Originally reported of Google Code with ID 698.

```
What software version are you running? Please provide the output of "gnt-
cluster --version", "gnt-cluster version", and "hspace --version".

root@vserver8:~# gnt-cluster  version
Software version: 2.9.2
Internode protocol: 2090000
Configuration format: 2090000
OS api version: 20
Export interface: 0
VCS version: v2.9.2

root@vserver8:~# gnt-cluster --version
gnt-cluster (ganeti v2.9.2) 2.9.2


<b>What distribution are you using?</b>

ubuntu 12.04.4 LTS

<b>What steps will reproduce the problem?</b>

moving an instance with "gnt-instance move..." is much slower than "moving" the instance with temporary disk template changed to drbd:

root@vserver8:~# time (gnt-instance modify -t drbd -n vserver8 instance1 ; gnt-instance failover -f instance1 ; gnt-instance modify -t plain instance1)
Thu Jan 30 08:59:19 2014 Converting template to drbd
Thu Jan 30 08:59:20 2014 Creating additional volumes...
Thu Jan 30 08:59:23 2014 Renaming original volumes...
Thu Jan 30 08:59:24 2014 Initializing DRBD devices...
Thu Jan 30 08:59:39 2014  - INFO: Waiting for instance instance1 to sync disks
Thu Jan 30 08:59:41 2014  - INFO: - device disk/1: 32.80% done, 16s remaining (estimated)
Thu Jan 30 08:59:41 2014  - INFO: - device disk/2:  2.40% done, 1m 25s remaining (estimated)
Thu Jan 30 09:00:42 2014  - INFO: Instance instance1's disks are in sync
Modified instance instance1
 - disk_template -> drbd
Please don't forget that most parameters take effect only at the next (re)start of the instance initiated by ganeti; restarting from within the instance will not be enough.
Thu Jan 30 09:00:43 2014  - INFO: Not checking memory on the secondary node as instance will not be started
Thu Jan 30 09:00:43 2014 Failover instance instance1
Thu Jan 30 09:00:43 2014 * not checking disk consistency as instance is not running
Thu Jan 30 09:00:43 2014 * shutting down instance on source node
Thu Jan 30 09:00:43 2014 * deactivating the instance's disks on source node
Thu Jan 30 09:00:47 2014 Converting template to plain
Thu Jan 30 09:00:47 2014 Removing volumes on the secondary node...
Thu Jan 30 09:00:48 2014 Removing unneeded volumes on the primary node...
Modified instance instance1
 - disk_template -> plain
Please don't forget that most parameters take effect only at the next (re)start of the instance initiated by ganeti; restarting from within the instance will not be enough.

real    1m31.299s
user    0m0.604s
sys     0m0.136s




root@vserver8:~# time gnt-instance move -f -n vserver9 instance1
Thu Jan 30 09:01:01 2014  - INFO: Not checking memory on the secondary node as instance will not be started
Thu Jan 30 09:01:01 2014  - INFO: Shutting down instance instance1 on source node vserver8
Thu Jan 30 09:01:03 2014  - INFO: Copying data for disk 0
Thu Jan 30 09:01:35 2014  - INFO: Copying data for disk 1
Thu Jan 30 09:02:53 2014  - INFO: Copying data for disk 2
Thu Jan 30 09:08:04 2014  - INFO: Removing the disks on the original node

real    7m4.380s
user    0m0.212s
sys     0m0.040s

i've tried it serval times in both directions and it is alwas the same. using disk template drbd for the "move" action is much faster.

maybe the normal move could be improved in some way? i am not sure why a normal move (dd?) is so much slower than drbd......



```
Originally added on 2014-01-30 08:15:29 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
-- Empty comment --
```
Originally added on 2014-01-30 08:41:43 +0000 UTC.

Added Labels: Type-Enhancement
Changed State: Accepted
Added to Milestone: Unplanned

**Comment by Ganeti-Issues-Migrator** [comment]
```
gnt-instance move performs a dd through ssh, so the data is sent encrypted over the wire. DRBD won't encrypt the data for you, so that might be part of the reason.

Please note that there are improvements in the move code in 2.10, but the fundamental difference (encrypted vs. unencrypted) will be the same for 2.10.
```
Originally added on 2014-01-30 15:35:16 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
FYI, the --compress option might help make things ever-so-slightly faster.
```
Originally added on 2014-01-30 15:39:19 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
I fear there is no support for the --compress option for intra-cluster moves in 2.9. But AFAIK it is supported as of 2.11 (sorry, correcting myself, this hasn't gotten in 2.10...).
```
Originally added on 2014-01-30 15:54:27 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
Oh, sorry, my bad indeed :)
```
Originally added on 2014-01-30 15:59:07 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
after doing some tests it looks like compression does not bring a big speedup.

but using an other cipher does:


root@vserver8:~# dd if=/dev/data/b0124bae-0b1f-4d99-9611-03905033e8d6.disk2_data | pv -s 8G | ssh vserver9 "cat > /dev/data/test"
16777216+0 Datensätze ein [====================================================================================================================================================================================================>  ] 99% ETA 0:00:00
16777216+0 Datensätze aus
8589934592 Bytes (8,6 GB) kopiert, 227,58 s, 37,7 MB/s
   8GB 0:03:47 [  36MB/s] [=====================================================================================================================================================================================================>] 100%


root@vserver8:~# dd if=/dev/data/b0124bae-0b1f-4d99-9611-03905033e8d6.disk2_data | pv -s 8G | ssh -C vserver9 "cat > /dev/data/test"
16777216+0 Datensätze ein [====================================================================================================================================================================================================>  ] 99% ETA 0:00:00
16777216+0 Datensätze aus
8589934592 Bytes (8,6 GB) kopiert, 207,34 s, 41,4 MB/s
   8GB 0:03:27 [39,5MB/s] [=====================================================================================================================================================================================================>] 100%


root@vserver8:~# dd if=/dev/data/b0124bae-0b1f-4d99-9611-03905033e8d6.disk2_data | pv -s 8G | ssh -c arcfour vserver9 "cat > /dev/data/test"
16777216+0 Datensätze ein [====================================================================================================================================================================================================>  ] 99% ETA 0:00:00
16777216+0 Datensätze aus
8589934592 Bytes (8,6 GB) kopiert, 128,764 s, 66,7 MB/s
   8GB 0:02:08 [63,6MB/s] [=====================================================================================================================================================================================================>] 100%

```
Originally added on 2014-01-31 07:16:18 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
It's quite hard to find the best compromise when doing moves. Things to consider are (at least):
 * network bandwidth
 * compute resources (for compression and encryption)
 * state of the disk (almost empty with a lot of empty blocks vs. almost all zeros on the disk)

There went some research into this in issue 473. Especially for intra-cluster moves I think we might go towards a more flexible move policy, where the cluster administrator should be able to select whether to compress or encrypt the data during moves, and if so which encryption should be used. Also, zeroing free space on disks before doing the move should be configurable on a per-cluster level (although I don't think it makes sense to zero free space for moves inside a cluster, as the network speed is typically higher than the zeroing + compressing speed).

We're open for any input, but there is currently quite some work being done in the inter-cluster move case which can be used in the intra-cluster move as well (but this will not be available before 2.12).

```
Originally added on 2014-01-31 15:18:57 +0000 UTC.

Mentioned Issue with GoogleCode ID 473 has been migrated to GitHub with ID #529.

**Comment by Ganeti-Issues-Migrator** [comment]
```
A lot of work went into improvimg instance move performance in newer versions of Ganeti (in particular 2.15, also 2.16 (to be released soon)). Feel free to file a new issue, if the situation has not improved in these newer versions.
```
Originally added on 2015-10-07 13:39:56 +0000 UTC.

Changed State: Released
